src/mini-batch.jl

```diff
@@ -0,0 +1,73 @@
+"""
+    MiniBatch(b::Int)
+
+"""
+struct MiniBatch <: AbstractKMeansAlg
+    b::Int  # batch size
+end
+
+
+MiniBatch() = MiniBatch(100)
+
+function km(algo::MiniBatch, X, k;
+                weights = nothing, metric = Euclidean(), n_threads = Threads.nthreads(),
+                k_init = "k-means++", init = nothing, max_iters = 300,
+                tol = 0, verbose = false, rng = Random.GLOBAL_RNG)
+
+    # Step 1. Select sample from X as specified by batch_size and weights
+    nrow, ncol = size(X)  # n_features, m_examples
+    T = eltype(X)
+    rand_idx = isnothing(weights) ? rand(rng, 1:ncol, algo.b) : wsample(rng, 1:ncol, weights, algo.b)
+    initial_batch_sample = X[:, rand_idx]
+
+    # Step 2. Initiate cluster centers from the initial sample
+    centroids = isnothing(init) ? smart_init(initial_batch_sample, k, n_threads, weights, rng, init=k_init).centroids : deepcopy(init)
+
+    # initialize nearest centers
+    labels = Vector{Int}(undef, algo.b)
+
+    # initialize no. of data in each cluster
+    N = zero(T)
+
+    # TODO: Initialise caches.(Dimensions right?)
+    centroids_cnt = Vector{Int}(undef, k)
+    centroids_new = zeros(T, nrow, k)
+
+    # TODO: Main Steps. Batch update centroids until convergence
+    for iter = 1:max_iters
+
+        # b examples picked randomly from X (Stage 5 in paper)
+        batch_sample = X[:, rand_idx]
+
+        # Cache/label the batch samples nearest to the centers
+        @inbounds for i in axes(batch_sample, 2)
+            min_dist = distance(metric, batch_sample, centroids, i, 1)
+            label = 1
+
+            for j in 2:size(centroids, 2)
+                dist = distance(metric, batch_sample, centroids, i, j)
+                label = dist < min_dist ? j : label
+                min_dist = dist < min_dist ? dist : min_dist
+            end
+
+            labels[i] = label
+        end
+
+        # TODO: Batch gradient step. Replace `nothing`s
+        for m = 1:algo.b
+            # TODO: Get cached center for this x
+            centroids_new = nothing # just a holder until I figure out
+
+            # TODO: Update per-center counts
+            N = centroids_cnt[labels[m]]
```

Comment:
You should update `centroids_cnt`, in paper: `c[x] = c[x] + 1` or something like that.